Add light support to Velbus integration

brefra opened the pull request:
Expose Velbus dimming modules as light entities

## Breaking Change:
none

## Description:

Add Light Entity control for Velbus modules that have a dimming feature and supported bij [Python-Velbus](https://pypi.org/project/python-velbus) library

**Related issue (if applicable):** none

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#11526

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

brefra committed:
Add light support to Velbus integration

Expose Velbus dimming modules as light entities

brefra committed:
Add light support to Velbus integration

Expose Velbus dimming modules as light entities

P-Verbrugge committed:
Updated formatting of total_blocks value (#30170)

The number of total blocks is always a round number. There can't be .1 or .11 blocks for example. The output is now always formatted with two decimals that are always 00.

heckler committed:
New date_time_utc display option added to the time_date sensor platform (#30158)

brefra committed:
Apply isort to imports of light.py

brefra committed:
Apply isort to imports of light.py

alandtse committed:
Enable config flow for Tesla (#28744)

* build: bump teslajsonpy to 0.2.0

* Remove tests

* feat: add config flow

* feat: add async

* perf: convert unnecessary async calls to sync

* feat: add charger voltage and current sensor

* feat: add options flow

* build: bump teslajsonpy to 0.2.0

* Remove icon property

* Revert climate mode change

* Remove charger sensor

* Simplify async_setup_platform

* Update homeassistant/components/tesla/sensor.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* Update homeassistant/components/tesla/binary_sensor.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

* Address requested changes

* Fix pylint error

* Address requested changes

* Update codeowners

* Fix pylint error

* Address requested changes

* Address requested change

* Remove unnecessary check for existing config entry

* Load scan_interval in async_setup_entry

* Include coverage of config_flow

* Add tests for full coverage

* Address requested test changes

* Remove unnecessary init lines

* Remove unnecessary init

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

dmulcahey committed:
Protect against bad data stored in ZHA (#30183)

homeassistant committed:
[ci skip] Translation update

Cereal2nd commented:
add the homeassistant/components/velbus/light.py to the .covergaerc file

cereal2nd reviewed (commented):
looks ok to

brefra committed:
Add Velbus light.py to .coveragerc

brefra committed:
Add Velbus light.py to .coveragerc

fabaff committed:
Upgrade zeroconf to 0.24.3 (#30187)

Quentame committed:
Fix AdGuard Home safe search sensor name (#30171)

* Fix AdGuard Home safe search sensor name

* Fix typo

garbled1 committed:
Add onewire devices and owserver remote host support (#29948)

* Add support for HobbyBoards sensors: Hub, Humidity, Moisture, PulseCounter
Add support for an owserver running on a remote host

* Cleanup and fixes for style/etc

* Forgot to convert to f-strings.

* Update with changes requested by @MartinHjelmare

* Pylint

* Fix to None

* Reverse logic for hobbyboard test to eliminate pylint complaint

* Modified manifest to list myself as codeowner

* Move some functions down into OneWireProxy instead of the top OneWire

* Oops, missed some decode()'s

* And another!

springstan committed:
Bump importlib-metadata to 1.3.0 (#30196)

* Bump importlib-metadata to 1.3.0

* Include setup.py and package_constraints.txt

springstan committed:
Bump datapoint to 0.9.5 (#30185)

ochlocracy committed:
Implement EqualizerController in Alexa for media_player. (#30159)

homeassistant committed:
[ci skip] Translation update

BKPepe committed:
Upgrade youtube_dl to 2019.12.25 (#30203)

ndonegan committed:
Pull track position from MPD status (#28579)

* Pull track position from MPD status()

This allows the progress bar to work when using the media-control card in lovelace.

* Actually commit flake8 fix!

* Extra documentation.

Mainly to trigger CI rerun.

* Updated to use self._media_position

rhadamantys committed:
Add support for enocean window handle FA 10 00 (Hoppe) (#29968)

* Added support for Somfy RTS wireless power socket and
Somfy Temperature Sensore Thermos Wirefree io

* Added code formatting fixes for commit 5faaf9c

* added support for RollerShutterRTSComponent from Somfy

* Added support for RTS roller shutter in set_cover_position

* added support for enocean window handle FA 10 00 (Hoppe)

* removed spaces in empty lines

* removal of rawdata attribute / code style

* isort fix

* PyLint fixes

* Improvements after review suggestions: rename device class to sensor type to avoid ambiguousness

* added import for DEVICE_CLASS_POWER from const.py

* removed window handle unit

HarlemSquirrel committed:
Update Vivotek camera component (#30191)

- Bump libpyvivotek version to 0.4.0
- Add digest authentication

scop committed:
Make Huawei LTE notify service name configurable (#30208)

* Default Huawei LTE notify service name to notify.huawei_lte, make configurable

Closes https://github.com/home-assistant/home-assistant/issues/29409

* Set default notify service name for uninvoked options flow

scop committed:
Add Huawei LTE integration suspend and resume services (#30207)

Useful e.g. if accessing the router web interface from another source
such as a web browser is temporarily required.

homeassistant committed:
[ci skip] Translation update

alandtse committed:
Bump teslajsonpy to 0.2.1 (#30217)

Closes #29922

PaulAnnekov committed:
bump tuyaha 0.0.5 (#30213)

bendavid committed:
Fix handling of symlinked device descriptors in keyboard_remote and move remaining sync io to executor thread pool (#30206)

* fix handling of symlinked device decriptors

* make check for symlinked paths more efficient

* make variable names pylint compliant

* move sync io during setup and device connect/disconnect to executor thread pool

* move remaining sync io during setup to executor thread pool

* remove unnecessary lambda functions

dmulcahey committed:
Bump ZHA dependencies (#30228)

* bump dependencies
* requirement files
* bump quirks to 0.0.30

homeassistant committed:
[ci skip] Translation update

frenck committed:
Bump shodan to 1.21.1 (#30234)

Kerwood committed:
DECONZ - Added support for Aqara single switch WXKG03LM (#30240)

homeassistant committed:
[ci skip] Translation update

springstan committed:
Bump python-qbittorrent to 0.4.1 (#30239)

tchellomello committed:
Bump ring to 0.2.8 to fix Oauth issues (#30245)

scop committed:
Fix Huawei LTE error message on service call without URL and routers (#30250)

Commit:
Bump dependency for HomematicIp cloud (#30237)

* Bump dependency for HomematicIp cloud

* Update test_data

xlfe committed:
Bump libpurecool to 0.6.0 (#30247)

* bump libpurecool to 0.6.0 - fixes home-assistant/home-assistant#26367

* update manifest too

scop committed:
Huawei LTE: Fix YAML options overriding ones set from GUI (#30249)

Closes https://github.com/home-assistant/home-assistant/issues/30221

Commit:
Add HmIP-MOD_TM to HomematicIP Cloud (#30255)

scop committed:
Note Huawei LTE notify service change requires restart (#30223)

Refs https://github.com/home-assistant/home-assistant/issues/30222

cereal2nd committed:
Bump python-velbus library to 2.0.32 to fix problems with the glaspanels (#30257)

balloob committed:
Whitelist Android/iOS auth callbacks (#30082)

* Whitelist Android/iOS

* Add iOS alternate flavor URLs

* Update indieauth.py

Co-authored-by: Robbie Trencheny <me@robbiet.us>

vangorra committed:
Fixing timezone issue which caused wrong selection of data to be used. (#30011)

Adminiuga committed:
Refactor ZHA channel logging (#30259)

Add channel.id property -- id unique for this the device only.

homeassistant committed:
[ci skip] Translation update

liollury committed:
Fix ble_tracker randomly pygatt thrown error (#28671)

* fix(ble_tracker): catch randomly pygatt thrown error

* fix(ble_tracker): merge except errors

Cereal2nd commented:
@brefra you will need to check all todo items in the PR before people will look at it

MartinHjelmare commented on homeassistant/components/velbus/light.py:
```diff
@@ -0,0 +1,99 @@
+        """Return the brightness of the light."""
+        return self._module.get_dimmer_state(self._channel)
+
+    def turn_on(self, **kwargs):
```
If the library supports asyncio we should use the home assistant async api, ie `async def async_turn_on`.

MartinHjelmare commented on homeassistant/components/velbus/light.py:
```diff
@@ -0,0 +1,99 @@
+        if ATTR_BRIGHTNESS in kwargs:
+            brightness = int(kwargs[ATTR_BRIGHTNESS])
+        else:
+            brightness = 0
```
Brightness isn't used in this case it seems.

brefra replied:
Fixed in [this commit](https://github.com/home-assistant/home-assistant/pull/30184/commits/0ab1b5bb6f0bccf6004945d541ea632408131f26)

MartinHjelmare commented on homeassistant/components/velbus/light.py:
```diff
@@ -0,0 +1,99 @@
+                    self._channel,
+                    brightness,
+                    transitiontime,
+                    self.async_schedule_update_ha_state(False),
```
Why are we calling `async_schedule_update_ha_state` here and passing the return value, `None`, to this method?

brefra replied:
You're right the callback is not required at all (a "left over" from early code testing)
I removed it in [this commit](https://github.com/home-assistant/home-assistant/pull/30184/commits/0ab1b5bb6f0bccf6004945d541ea632408131f26)

MartinHjelmare commented:
Off topic:
@Cereal2nd
It would be good to fix this use of async api inside sync context:
https://github.com/home-assistant/home-assistant/blob/a91b0058229974aed578fb20e833df4454d007f9/homeassistant/components/velbus/__init__.py#L70-L72

Cereal2nd commented:
@MartinHjelmare 
what exactly do we need to change?
I'm not 100% how all this asyncio works exactly

MartinHjelmare commented:
I think we need to use `asyncio.run_coroutine_threadsafe` to schedule the forwarding coroutine on the hass event loop.

https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe

```py
coro = hass.config_entries.async_forward_entry_setup(entry, category)
asyncio.run_coroutine_threadsafe(coro, hass.loop)
```

mockersf committed:
Fix creating smappee sensors when remote is not active (#30270)

scop committed:
Upgrade rflink to 0.0.50, ignore_devices now supports * and ? anywhere (#30268)

https://github.com/aequitas/python-rflink/releases/tag/0.0.50
https://github.com/aequitas/python-rflink/releases/tag/0.0.49
https://github.com/aequitas/python-rflink/releases/tag/0.0.48
https://github.com/aequitas/python-rflink/releases/tag/0.0.47

brefra committed:
Removed callbacks for updates

brefra committed:
Removed callbacks for updates

homeassistant committed:
[ci skip] Translation update

Jc2k committed:
Accept homekit_controller pairing codes both with and without dashes (#30273)

* Handle MalformedPinError from homekit_python
* Handle both formats of pin codes

Jc2k committed:
Add homekit_controller service.sensor.smoke (#30269)

brefra committed:
Applied black formatting

brefra committed:
Applied black formatting

fabaff committed:
Upgrade beautifulsoup4 to 4.8.2 (#30274)

fabaff committed:
Add custom validator for countries (#30280)

alandtse committed:
Change default icons for Tesla components (#30288)

andygrunwald committed:
Contributing: Add note about feature suggestions and bug tracking (#30225)

MartinHjelmare commented on homeassistant/components/velbus/light.py:
```diff
@@ -0,0 +1,85 @@
+
+    def turn_off(self, **kwargs):
+        """Instruct the velbus light to turn off."""
+        if ATTR_BRIGHTNESS in kwargs:
```
Brightness is not valid for turn off.

brefra replied:
Thanks I didn't know that, but it sounds very logical.
Fixed in [this ](https://github.com/home-assistant/home-assistant/pull/30184/commits/711d9ef5dc88f6b8310445004b7174547e544ce2) additional commit

taisholland committed:
Bump ihcsdk to 2.4.0 (#30279)

* Update requirements_all.txt

update ihcsdk to version 2.4.0

* Update manifest.json

upgrade to version 2.4.0 of ihcsdk

Kane610 committed:
Handle wired bug on restart (#30276)

brefra committed:
Removed brightness in turn_off

Brightness is not a valid argument when turning off lights

brefra committed:
Removed brightness in turn_off

Brightness is not a valid argument when turning off lights

brefra committed:
Apply black formatting

brefra committed:
Apply black formatting

andrewsayre committed:
Bump pysmartthings 0.7.0 (#30302)

MartinHjelmare commented on homeassistant/components/velbus/light.py:
```diff
@@ -0,0 +1,82 @@
+        try:
+            if ATTR_BRIGHTNESS in kwargs:
+                self._module.set_dimmer_state(
+                    self._channel, int(kwargs[ATTR_BRIGHTNESS]), transitiontime,
```
Brightness is already validated as an integer by the service schema. We don't need to copy to integer.

brefra replied:
The same applies to `ATTR_TRANSITION`
Both fixed in [this](https://github.com/home-assistant/home-assistant/pull/30184/commits/41a319f2011d4ff77c69b2af3b8f3e2b2c65e2b6) commit

brefra committed:
Removed integer conversions

and duplicate turn off call

brefra committed:
Removed integer conversions

and duplicate turn off call

brefra committed:
Applied pylint rules

brefra committed:
Applied pylint rules

brefra committed:
Apply black formating

brefra committed:
Apply black formating

MartinHjelmare reviewed (approved):
Looks good!

homeassistant committed:
[ci skip] Translation update

pnbruckner committed:
Ignore google_maps updates when last_seen goes backwards (#30178)

MartinHjelmare commented:
The automated PR checks are missing except the CLA check. I feel a bit uneasy merging this without those checks. I'm not sure why the checks are gone.

Can we rebase this on latest dev to see if we get the checks back?

bieniu committed:
Add GIOS integration (#28719)

* Initial commit

* Add gios to requirements

* Add tests

* Update .coveragerc file

* Run gen_requirements_all.py

* Change DEFAULT_SCAN_INTERVAL

* Better strings

* Bump library version

* run script.hassfest

* run isort

* Add icons mapping

* Remove unnecessary f-string

* Remove unnecessary listener

* Refactoring config_flow

* Add unique_id to config entry

* Change AQI states to consts in English

* Remove unused init

* Remove unused exception

* Remove private instance attribute

* Remove overwrite state property

* Fix pylint error

* Add SCAN_INTERVAL for air_quality entity

* Add _abort_if_unique_id_configured()

brefra committed:
Merge branch 'dev' of https://github.com/brefra/home-assistant into dev

brefra commented:
I messed up this PR trying to rebase it. Will submit new PR
